src/wp-includes/class-wp-query.php

```diff
@@ -1781,7 +1781,9 @@ public function get_posts() {
 		 *
 		 * @param WP_Query $query The WP_Query instance (passed by reference).
 		 */
-		do_action_ref_array( 'pre_get_posts', array( &$this ) );
+		if ( ! did_action( 'pre_get_posts' ) ) {
```

Comment:
```suggestion
		if ( ! doing_action( 'pre_get_posts' ) ) {
```

I think `doing_action()` is a better match for your intention here, to prevent the action running if it is already doing so.

Using `did_action()` would stop the action running on all but the first query on a page, this may end up exposing data to users a plugin is trying to keep private.

Thanks for the pull request.